### DISCUSSION TITLE
SPI 12-bit, sending data

### DISCUSSION THREAD

**Discussion by Davis-Garrad** [micropython/micropython#15888] [General]
Hello,

As the title says, I've configured my Raspberry Pi Pico's SPI bus for 12-bit interactions through `init`. This is fine - I haven't yet checked if my pico can natively support this, but right now I'm only really concerned about the code... How, in Python, can I send 12 bits, when the argument to `write` must be a bytearray? I couldn't find anything on the docs about this.

The first idea that came to mind was that, perhaps it was just that two bytes need to be provided to the function... But then I questioned which order the 12 bits would be read in. With some fiddling, I can certainly figure it out if this is the case, but I'm no Python pro and this seems to be a hole in the documentation, so I figure I'll ask. If I find an answer myself, I'll document here for future use.

Thanks!

**Comment by peterhinch** [comment]
What hardware are you driving? The only chips I'm aware of that do 12-bit transfers are display drivers. In this case you are normally sending a whole buffer full of data; you can pack the 12-bit pixels into bytearrays and use normal transfers. Then every pair of pixels occupies three bytes. You may enjoy the challenge of figuring out how to code this. Otherwise give us more details of your setup and I'm sure we can help.

**Comment by Davis-Garrad** [comment]
Ah, you're totally right - it's sent/received in larger, 8-aligned groups,
but parsed in 12s. Seems relatively straightforward to implement (or
certainly moreso than trying to make python parse nibbles)... Thanks for
your help!

**Comment by bidrohini** [comment]
Since a byte is 8 bits, 12 bits can be split into:
The first byte contains the higher 8 bits of the 12-bit number.
The second byte contains the lower 4 bits (shifted into the upper half of the byte, with the lower half padded with zeros).
